### ISSUE TITLE
Simplify caching support

### ISSUE THREAD

**Issue by stas** [fast-jsonapi/fast_jsonapi#51]
## Expected Behavior

Basically, the way caching support was designed is not very flexible and very Rails-specific.

Particularly we should add:
 * [ ] support for versioning
 * [ ] support for namespacing
 * [ ] less rails-specific API calls
 * [ ] unified/merged caching options

## Actual Behavior

We implemented only a small subset of the required caching functionality and it's all very Rails-specific. Hence the #49 and #50 exist...

**Comment by stas** [comment]
A first step would be to get rid of the custom settings and allow our DSL to accept a bunch of options we can send over directly to the caching method call.

In rails these are available, we should leverage them:
https://github.com/rails/rails/blob/80e72c5eb7d4f79cc8eaa63237e5a60dff48f4f4/activesupport/lib/active_support/cache.rb#L22-L24

**Comment by doits** [comment]
Makes sense to allow to configure all available caching options and pass them directly.

Doesn't it conflict with "less rails-specific API calls" then, though? At least if options are passed as is to the cache implementation. Or should there be a wrapper that translates fastjson-cache-options to actuall caching implementation options?

**Comment by stas** [comment]
> Doesn't it conflict with "less rails-specific API calls" then, though? At least if options are passed as is to the cache implementation. Or should there be a wrapper that translates fastjson-cache-options to actuall caching implementation options?

If we assume that the users will provide their own cache store (or default it to the `Rails.cache`), there should be no conflicts. In the end it's everyone's own configuration, so if it's not working, it's in the _userland_ not because of the library.

I'll quote the example I left in #50:
```ruby
class MySerializer
  include FastJsonApi::ObjectSerializer

  cache_options(store: ActiveSupport::Cache::MemoryStore.new, ...)
  # Or default it to rails if availble
  cache_options(store: Rails.cache, ...)
end
```

**Comment by doits** [comment]
OK, and in this case the options can be simply passed directly but the ones that change how the methods are called.

```rb
cache_options(
  store: ActiveSupport::Cache::MemoryStore.new(namespace: 'whatever'),
  versioning: true # <-- still needed here, at least if 
                   #     the caching key for the record is 
                   #     derived by fast jsonapi like now
)
```

**Comment by stas** [comment]
@doits namespace is supported on `fetch` or `write`, it can stay separate from the cache store intialization.

Example:
```
irb(main):020:0> Rails.cache.clear
=> 0
irb(main):021:0> Rails.cache.write(:a, 'value_for_a', namespace: :v1)
=> true
irb(main):022:0> Rails.cache.fetch(:a)
=> nil
irb(main):023:0> Rails.cache.fetch('v1:a')
=> "value_for_a"
irb(main):025:0> Rails.cache.fetch(:a, namespace: :v1)
=> "value_for_a"
```

Unfortunately the Rails documentation doesn't reflect all the API functionality.

**Comment by stas** [comment]
May I ask you exactly, how do you expect this versioning to work?

I'm asking because the AS caching already does some versioning support:
https://github.com/rails/rails/blob/80e72c5eb7d4f79cc8eaa63237e5a60dff48f4f4/activerecord/lib/active_record/integration.rb#L115 

Example:
```ruby
> u1 = Fabricate :user, email: 's@s.co'
...
> u1.cache_key
=> "users/e172f973-b74e-4d3f-a087-522a68ca25ce"
> u1.cache_key_with_version
=> "users/e172f973-b74e-4d3f-a087-522a68ca25ce-20200120135450435358"
> u1.cache_version
=> "20200120135450435358"
```

**Comment by doits** [comment]
@stas exactly, it is about *supporting* the new cache versioning behavior! It is *not* about implementing a different one.

Because `.cache_key` does *not* change if record is updated. But current caching implementation in fast jsonpi uses this as the only key to fetch the cache and it will fetch outdated records by this.

So for example in current fast jsonapi cache implementation it will do this:

```rb
> User.cache_versioning
=> true # cache versioning is enabled

> UserSerializer.cache_options(enabled: true) # enable caching

> u1.cache_key 
=> "users/e172f973-b74e-4d3f-a087-522a68ca25ce"

> u1.cache_version
=> "20200120135450435358"

> UserSerializer.new(u1).serializable_hash
=> { ..., attributes: { name: 'My user name' } }

> u1.update name: 'Changed name'
=> true

> u1.cache_key 
=> "users/e172f973-b74e-4d3f-a087-522a68ca25ce" # <-- still the same cache key

> u1.cache_version
=> "2020012014225043442" # <-- different cache version

> UserSerializer.new(u1).serializable_hash
=> { ..., attributes: { name: 'My user name' } } # <-- old name, even though record was updated
```

This happens because fast jsonapi does a `Rails.cache.fetch(u.cache_key)`. It should instead do a `Rails.cache.fetch(u1.cache_key, version: u1.cache_version)` in this case.

So how does fast jsonpi know whether `version: u1.cache_version` should be used or not? It could infer it automatically, e.g. check `User.cache_versioning` - but this is Rails specific and therefore maybe not wanted. Or it could be explicit know because it is passed as an option:

```rb
cache_options(versioning: true)
```

**Comment by stas** [comment]
I see... it's just another Rails-esque trick.

Basically, if we send the object/model to the cache store, it will automatically figure out what to call, either `cache_key` or `cache_key_with_version`. So I assume we should make sure the whole record is passed and not explicitly the `cache_key`.

Example:
```ruby
> u1 = User.first

> u1.cache_versioning
=> false

> u1.cache_key_with_version
=> "users/e172f973-b74e-4d3f-a087-522a68ca25ce-20200120135450435358"

> Rails.cache.write(u1, u1.id)
=> true

> Rails.cache.as_json
=> {"options"=>{}, "data"=>{"users/e172f973-b74e-4d3f-a087-522a68ca25ce-20200120135450435358"=>{"value"=>"e172f973-b74e-4d3f-a087-522a68ca25ce", "version"=>nil, "created_at"=>1579532862.2533202, "expires_in"=>nil}}, "key_access"=>{"users/e172f973-b74e-4d3f-a087-522a68ca25ce-20200120135450435358"=>1579532862.2536123}, "max_size"=>33554432, "max_prune_time"=>2, "cache_size"=>339, "monitor"=>{}, "pruning"=>false}

```

**Comment by doits** [comment]
> Basically, if we send the object/model to the cache store, it will automatically figure out what to call, either cache_key or cache_key_with_version. So I assume we should make sure the whole record is passed and not explicitly the cache_key.

Yes, this is a solution, too. This means that *not* fast jsonapi decides about the correct key to use but the cache store can decide how to derive the correct cache key (or version) from the record.

**Comment by stas** [comment]
I'm ok with that. I'll be happy to review any PRs related to that if you decide to help with it!

**Comment by doits** [comment]
@stas I gave it a shot in #52. Feel free to comment.

**Comment by stas** [comment]
Closed in #52
